### PR TITLE
shell: move away from accountsservice

### PR DESCRIPTION
**Pull request by fridex**
Proposing patches based on https://github.com/cockpit-project/cockpit/issues/1853

### CODE REVIEW

**Comment by stefwalter** [comment]
Wow. First of all I wanted to say "Awesome". Thanks for doing this.

Now a little review.

**Comment by stefwalter** [comment]
Fixes #1853

File: pkg/shell/cockpit-accounts.js [hunk 1]
```diff
@@ -159,6 +132,84 @@ function off_account_changes(client, id) {
     $(client).off("." + id);
 }
 
+function parse_passwd_content(content) {
+    var ret = [ ];
+    var lines = content.split('\n');
+    var column;
+
+    for (var i = 0; i < lines.length - 1 /* without blank line */; i++) {
```
Comment (stefwalter) [inline pkg/shell/cockpit-accounts.js]: I would suggest simply ignoring blank lines.

File: pkg/shell/cockpit-accounts.js [hunk 2]
```diff
@@ -159,6 +132,84 @@ function off_account_changes(client, id) {
...
+        column = lines[i].split(':');
+        ret[i] = [];
+        ret[i]["name"] = column[0];
+        ret[i]["password"] = column[1];
+        ret[i]["uid"] = column[2];
+        ret[i]["gid"] = column[3];
+        ret[i]["gecos"] = column[4];
+        ret[i]["home"] = column[5];
+        ret[i]["shell"] = column[6];
```
Comment (stefwalter) [inline pkg/shell/cockpit-accounts.js]: In your opinion, should we strip the line, or expect that there are no trailing characters such as ```\r``` ?

File: pkg/shell/cockpit-accounts.js [hunk 3]
```diff
@@ -487,54 +621,65 @@ PageAccount.prototype = {
...
     logout_account: function() {
-        var me = this;
-        this.account.call('KillSessions',
-                          function (error) {
-                              if (error) {
-                                  shell.show_unexpected_error(error);
-                                  me.update ();
-                              }
-                          });
+        cockpit.spawn(["/usr/bin/pkill", "-9", "-u", this.account["name"]], { "superuser": true})
```
Comment (stefwalter) [inline pkg/shell/cockpit-accounts.js]: We should be using ```loginctl kill-user``` and do this through logind.

File: tools/cockpit.spec [hunk 4]
```diff
@@ -103,8 +102,8 @@ Requires: realmd
 Requires: storaged
```
Comment (stefwalter) [inline tools/cockpit.spec]: ```%package shell```

Needs a dependency on expect and shadow-utils

**Comment by stefwalter** [comment]
screenshot from 2015-03-03 07 39 32

Either we should display the placeholder avatars (the gray "user" image) or we need to update the design so that it doesn't look like this.

**Comment by stefwalter** [comment]
screenshot from 2015-03-03 07 40 42

 * I think it's a good idea to display the real group names (perhaps in addition to the fancy names). But we would need to change the string 'Roles' here
 * There's too much space between the 'Login Shell' and 'Access' rows.
 * Need a better look for the '**Never logged in**', suggest italics, without the stars.

**Comment by stefwalter** [comment]
screenshot from 2015-03-03 07 43 00

This error happened when I clicked the 'Terminate Session' button for a user that was not logged in.

**Comment by stefwalter** [comment]
screenshot from 2015-03-03 07 43 42

We can't just display all the groups here, since most of them are not useful for users. I would suggest limiting this the same way it was limited before (wheel and docker groups) and opening a separate issue to figure out how to use additional groups here.

**Comment by stefwalter** [comment]
When I change the 'Real Name' things flicker very fast, and the 'User name' gets updated with the Real name I just typed for about half a second.

**Comment by stefwalter** [comment]
screenshot from 2015-03-03 07 50 47

The 'Last Login' field is not really readable. Should we try and parse it?

**Comment by stefwalter** [comment]
@andreasn There's lots of design questions here that need to be worked out.

**Comment by fridex** [comment]
Thanks for the review!

I am proposing patches which should fix most of the issues.

I have some additional questions/notes to discuss:

> We can't just display all the groups here, since most of them are not useful for users. I would suggest > limiting this the same way it was limited before (wheel and docker groups) and opening a separate > issue to figure out how to use additional groups here.
>
> Alternatively, we could let the user type other groups rather than making a confusing list.

I agree that there could be a lot of groups and it could be confusing for a user. On the other hand, it gives an admin a possibility to be more powerful without firing up a terminal. I can imagine a button there to expand list to a full group management ("Show all groups") and hide all other groups by
default. Is this feature acceptable or should there be left only docker groups with wheel?

> The 'Last Login' field is not really readable. Should we try and parse it?

It looks like there is a problem with locale - it should be executed with "LC_ALL=C" as well. There is no problem with parsing at all. What is expected to be shown? Could you provide me an example?

According to discussion in https://github.com/cockpit-project/cockpit/issues/1853#issuecomment-77003168 everything should be "real time" displayed. The only thing which is currently not "real time" is only "Last Login" user field. It could be easily monitored by watching /var/log/lastlog database. This could be easily implemented if desired.

**Comment by stefwalter** [comment]
> I agree that there could be a lot of groups and it could be confusing for a user. On the other hand, it gives an admin a possibility to be more powerful without firing up a terminal. I can imagine a button there to expand list to a full group management ("Show all groups") and hide all other groups by
default. Is this feature acceptable or should there be left only docker groups with wheel?

I think that if there is significant redesign, it should happen in separate pull requests and work. So I would suggest leaving the look for this pull request pretty much identical to how it was previously.

> It looks like there is a problem with locale - it should be executed with "LC_ALL=C" as well. There is no problem with parsing at all. What is expected to be shown? Could you provide me an example?

The way it looked before, if possible.

  Last Login 	15.1.2015 15:54:24

and

  Last Login 	Logged In

**Comment by stefwalter** [comment]
> I am proposing patches which should fix most of the issues.

Nice. Thanks!

**Comment by andreasn** [comment]
So I looked over this briefly, and @stefwalter have pointed out a lot of issues already, and I agree with those.
I was a bit surprised to see this pull request not only porting things over from accountservice, but also grow new things, such as the ability to set login shell. I would rather see that as a separate pull request.

**Comment by andreasn** [comment]
If this can be rebased upon latest master, that would be great. I've had some issues testing this properly.

**Comment by andreasn** [comment]
So it seems the Anaconda installer is using "Full Name" rather than "Real Name" (at least on RHEL 6.7). I think this is what's used across other tools as well and I think that would be a better string here too.

**Comment by fridex** [comment]
Rebased upon currently latest master, notes/issues have been fixed and merged to commits.
My notes:
  * @andreasn there was "Real Name" in the original code, could be adjusted if desired
  * According to https://github.com/cockpit-project/cockpit/issues/1853#issuecomment-77003168 changes in "Lock/Unlock" and in "Last Login" are not "real time". This could be implemented by watching /var/log/lastlog database and /etc/shadow for changes.

**Comment by stefwalter** [comment]
Thanks. I'll try and review and/or merge this while travelling today.

> According to #1853 (comment) changes in "Lock/Unlock" and in "Last Login" are not "real time". This could be implemented by watching /var/log/lastlog database and /etc/shadow for changes.

Fine for now. @fridex could you file some follow up issues so we don't forget about these?

**Comment by stefwalter** [comment]
Wow. this behaves really well and looks good.

A few little problems:

 * It seems that 'Real Name' and 'User name' in the 'Create New Account' dialog are mixed up.
 * When I create a new account, setting the password fails.

I think that I need to migrate cockpit-setup.js to avoid accountsservice.

@mvollmer, we need to build new images with 'expect' in them.

**Comment by fridex** [comment]
Yes, you are right. They were mixed - that's why setting password didn't work. Sorry for that.

**Comment by andreasn** [comment]
I get a merge conflict when I try to pull this now.

**Comment by fridex** [comment]
Rebased upon currently latest master.

**Comment by dperpeet** [comment]
This is also linked to #174

**Comment by andreasn** [comment]
I tried to create a new user, and I got an error that it "Failed to change password"
The password was "banana"

**Comment by andreasn** [comment]
The "lock account" doesn't seem work.

**Comment by fridex** [comment]
> The password was "banana"
This password is too weak. Try to change password using passwd (this is how it is currently implemented) - it will cause an error (same for palindromes).

> The "lock account" doesn't seem work.
Could you please be more specific with a scenario?

**Comment by andreasn** [comment]
> This password is too weak.

Ah. I see. In this case we would need an inline error, similar to the duplicate password one.
The message "Failed to create password" is too unspecific. I thought it was a backend error.

> The "lock account" doesn't seem work.

Trying to uncheck the checkbox just checks it back again.

**Comment by andreasn** [comment]
What kind of passwords work? I fail using pretty much everything.

**Comment by fridex** [comment]
> What kind of passwords work? I fail using pretty much everything.

It depends on your environment. I suppose you use pam_unix module (man pam_unix). This brings me to another issue. What if server is configured to use other auth methods (such as Radius or Kerberos)?

> Trying to uncheck the checkbox just checks it back again.

Cannot reproduce it. I'm using Chrome/Firefox on Fedora 21. I suppose it occurs when administrating an existing account.

**Comment by andreasn** [comment]
Can you post an example password that will certainly work?
Just so I can verify nothing is broken on my side

**Comment by fridex** [comment]
Try "nov6enynad" for example. It works for me.

**Comment by stefwalter** [comment]
> It depends on your environment. I suppose you use pam_unix module (man pam_unix). This brings me to another issue. What if server is configured to use other auth methods (such as Radius or Kerberos)?

We only manage local accounts in this interface here. There are other interfaces for kerberos and other accounts (such as IPA).

**Comment by stefwalter** [comment]
I'll do a follow up of this, and merge the parts that change the 'Administrator Accounts' component. I won't yet merge the removal of accountsservice dep and code from cockpitd, as that is still in use by the initial setup code.

**Comment by stefwalter** [comment]
I'm working on this in #1972

**Comment by stefwalter** [comment]
Remaining to be merged from this pull request: Deletion of cockpitd accountsservice code, and dependency.

**Pull request closed without merging**